source/dubregistry/internal/utils.d

```diff
@@ -13,3 +26,42 @@ string black(string url)
 @safe {
 	return black(URL(url)).toString();
 }
+
+/**
+ * Params:
+ *   file = the file to convert
+ *   deleteFinish = if true, delete the input file after a successful conversion
+ * Returns: the PNG stream of the icon or empty on failure
+ * Throws: Exception if name is empty or logo already exists and deleteExisting is not true
+ */
+bdata_t generateLogo(NativePath file) @safe
+{
+	static assert (isWeaklyIsolated!(typeof(&generateLogoUnsafe)));
+	static assert (isWeaklyIsolated!NativePath);
+	return (() @trusted => async(&generateLogoUnsafe, file).getResult())();
+}
+
+private bdata_t generateLogoUnsafe(NativePath file) @safe
+{
+	import std.array : appender;
+
+	auto png = pipeProcess(["convert", file.toNativeString, "-resize", "512x512>", "-"]);
+
+	if (png.pid.wait != 0)
+	{
+		(() @trusted {
+			foreach (error; png.stderr.byLine)
+				logDebug("convert error: %s", error);
+		})();
+		return bdata_t.init;
+	}
+
+	auto a = appender!(ubyte[])();
+
+	(() @trusted {
+		foreach (chunk; png.stdout.byChunk(4096))
+			a.put(chunk);
+	})();
+
+	return cast(bdata_t)a.data.idup;
```

Comment:
Appender uses the GC to allocate plus it's local, so you can use `assumeUnique` and avoid the idup copy

Reply:
Or change to `appender!(immutable(ubyte)[])`

Reply:
well, in this case you could simply use `execute` - this would save a few lines?
(I will look into making modifying `execute`)

Reply:
execute converts to a string, but the std output gets a binary png file.

Reply:
> I don't want to risk an eventual std.utf.validate getting in there.

FTR with `cast(immutable(ubyte)[]) execute(["convert", ...])` `validate` isn't called.
Appender uses `ElementEncodingType`, not `ElementType` to avoid auto-decoding.

Reply:
Just for completeness, there is `vibe.stream.operations.readAll` to read all of a stream into a buffer.

Reply:
does it work for std.stdio.File like you get when spawning a process?

Reply:
No, would have to be wrapped inside a `StdFileStream` for that, but that is rather ineffecient currently.